onnx/defs/tensor/defs.cc

```diff
@@ -3912,4 +3990,156 @@ ONNX_OPERATOR_SET_SCHEMA(
+          if (ctx.getNumInputs() != 2) {
+            fail_type_inference("CenterCropPad op must have 2 inputs.");
+          }
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          if (!hasNInputShapes(ctx, 1)) {
+            return;
+          }
+          // Shape Inference if shape is initializer
+          const TensorProto* cropShapeInitializer = ctx.getInputData(1);
+          if (!cropShapeInitializer) {
+            return;
+          }
+
+          // don't know data_type - can't proceed
+          if (!cropShapeInitializer->has_data_type())
+            return;
+
+          const auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
+          const int64_t input_rank = input_shape.dim_size();
+
+          std::vector<int64_t> shape;
+          if (cropShapeInitializer->data_type() == TensorProto::INT64) {
+            const auto& data = ParseData<int64_t>(cropShapeInitializer);
+            shape.insert(shape.end(), data.begin(), data.end());
+          } else if (cropShapeInitializer->data_type() == TensorProto::INT32) {
+            const auto& data = ParseData<int32_t>(cropShapeInitializer);
+            shape.insert(shape.end(), data.begin(), data.end());
+          } else {
+            // unaccepted data type
+            fail_shape_inference("`shape` only supports `int32_t` or `int64_t` inputs");
+          }
+
+          auto axes_attr = ctx.getAttribute("axes");
+          std::vector<int64_t> axes;
+          if (axes_attr) {
+            axes = RetrieveValues<int64_t>(*axes_attr);
+          } else {
+            axes.resize(input_rank);
+            std::iota(axes.begin(), axes.end(), 0);
+          }
+
+          if (shape.size() != axes.size()) {
+            fail_shape_inference(
+                "Number of elements of input 'shape' (", shape.size(),
+                ") does not match the number of axes (", axes.size(),").");
+          }
+
+          // Populating default dims
+          std::vector<TensorShapeProto_Dimension*> out_dims(input_rank);
+          auto* output_shape = getOutputShape(ctx, 0);
+          for (int i = 0; i < input_rank; ++i) {
+            out_dims[i] = output_shape->add_dim();
+            const auto& input_dim = input_shape.dim(i);
+            if (input_dim.has_dim_value()) {
+              out_dims[i]->set_dim_value(input_dim.dim_value());
+            } else if (input_dim.has_dim_param()) {
+              out_dims[i]->set_dim_param(input_dim.dim_param());
+            }
+          }
+          int j = 0;
+          for (int axis : axes) {
+            out_dims[axis]->set_dim_value(shape[j++]);
+          }
+        })
+        .SetContextDependentFunctionBodyBuilder(
+            [](const FunctionBodyBuildContext& ctx, const OpSchema& schema, FunctionProto& functionProto) {
+              const auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
+              const int64_t input_rank = input_shape.dim_size();
```

Comment:
Unfortunately, this is going to be a problem. The shape we can get is only a static-shape, not a dynamic-shape. Statically, we may not know the rank of the input.

Reply:
Instead, one alternative is to get the (full) shape, convert the axes attribute into a tensor, and use it to index into the full-shape using some form of gather op.